Implement a CombinedFileProvider

mgrosperrin opened the pull request:
This PR is for the issue #49 

The behavior for each method is:
1. `GetFileInfo`: returns the first `IFileInfo` of all provided `IFileProvider` where the `Exists` property is `true`,
2. `GetDirectoryContents`: merges the the contents of all provided `IFileProvider`. On duplicate entries (based on `Name` property), only the first  entry is kept,
3. `Watch`: returns a `IChangeToken` that merges the `IChangeToken` of all provided `IFileProvider`. When one of the token raises the change notification, the returned token raises it.

dnfclas commented:
Hi __@mgrosperrin__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

MartinJohns commented on global.json:
```diff
@@ -1,3 +1,6 @@
-    "projects": ["src"]
+    "projects": [ "src" ],
+    "sdk": {
+        "version": "1.0.0-beta8"
```
I don't think this modification should be part of the PR. Especially not since the latest stable version is RC1, not beta8.

MartinJohns commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,162 @@
+            return combinedDirectoryContents;
+        }
+
+        public IChangeToken Watch(string pattern)
```
No documentation for this method? I think especially for this one it is interesting.

mgrosperrin commented:
Hi @MartinJohns,
thanks for your remarks. I've updated the global.json to remove my changes.
I have added documentation to the  `Watch` method, and add more info on the behavior of the others methods as well. I have put it in the `returns` part of the documentation. Let me know if this is good, or if it will be better in a `remarks` element.

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+        /// <returns>The file information. Caller must check Exists property. This will be the first existing <see cref="IFileInfo"/> returned by the provided <see cref="IFileProvider"/> or a not found <see cref="IFileInfo"/> if no existing files is found.</returns>
+        public IFileInfo GetFileInfo(string subpath)
+        {
+            foreach (var fileProvider in _fileProviders)
```
nit: you could change `_fileProviders`'s type to `IFileProvider[]` to avoid allocating an enumerator here (you don't have to explicitly use a `for` loop, since the compiler is supposed to optimize that for you).

mgrosperrin replied:
I don't understand your point: `IFileProvider[]` implements `IEnumerable<IFileProvider>` so when an enumerator will be allocated (which will not be in your case) ?

PinpointTownes replied:
When a variable is declared as an array, the compiler automatically replaces the `foreach` loop by a `for` loop and directly accesses the items via `array[index]` without having to instantiate a new enumerator. The compiler can't apply the same optimization when the type is `IEnumerable<>`.

mgrosperrin replied:
Thanks, I'm happy to learn something today!

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            // gets the content of all directories and merged them
+            var existingDirectoryContentsForAllProviders = _fileProviders.Select(
```
nit: for maximal performance, you could avoid using LINQ here.

mgrosperrin replied:
OK, I will change it for a ~~old-school~~ traditionnal loop :cry:

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+        {
+            private readonly IEnumerable<IChangeToken> _changeTokens;
+
+            public CombinedFileChangeToken(IEnumerable<IChangeToken> changeTokens)
```
Same remark as above: since this parameter is always used with a list, you could replace IEnumerable with a list or an array.

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+
+        private class CombinedDirectoryContents : IDirectoryContents
+        {
+            private readonly Dictionary<string, IFileInfo> _files = new Dictionary<string, IFileInfo>();
```
Curious: why using a dictionary here? AFAICT, the key doesn't seem to be used anywhere.

mgrosperrin replied:
The dictionary is used to filter duplicate `IFileInfo` (see line 102). I can rewrite it to use a `Hash<string>` but I will have the hashset and the list of corresponding `IFileInfo`. I think it was simpler to do it this way but I can change.

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+
+            private class Disposables : IDisposable
+            {
+                private readonly IEnumerable<IDisposable> _disposables;
```
nit: same remark :smile:

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+        /// Enumerate a directory at the given path, if any.
+        /// </summary>
+        /// <param name="subpath">The path that identifies the directory</param>
+        /// <returns>Contents of the directory. Caller must check Exists property. The content is a merge of the contents of the provided <see cref="IFileProvider"/>. When there is multiple <see cref="IFileInfo"/> with the same Name property, only the first one is included on the results.</returns>
```
nit: this line seems a bit long :smile:

mgrosperrin replied:
Yeah, it's the problem when VS is configured to do "Word wrap". I will add some new lines :smile:

PinpointTownes commented:
Definitely a must have! :smile:

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,168 @@
+
+            public CombinedDirectoryContents(IEnumerable<IEnumerable<IFileInfo>> listOfFiles)
+            {
+                foreach (var files in listOfFiles)
```
This code path will immediately enumerate all the files exposed by the different providers, which may take some time. You should maybe consider deferring this operation to the last moment (in `GetEnumerator`) and using an iterator.

mgrosperrin replied:
I have made the assumption that the directory contents will be pre-calculated (based on the two other providers), but you're right. I can move it to a `Lazy<IEnumerable<IFileInfo>>` (to avoiding dealing with the thread-safe and evaluating only once issues).

davidfowl commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,193 @@
+        /// Initializes a new instance of the <see cref="CombinedFileProvider" /> class using a list of file provider.
+        /// </summary>
+        /// <param name="fileProviders"></param>
+        public CombinedFileProvider(params IFileProvider[] fileProviders)
```
I'd avoid params. Just take an `IEnumerable<IFileProvider>`

PinpointTownes replied:
A `params` overload definitely makes sense, because that's how people will use this combined thingy: 

```csharp
options.FileProvider = new CombinedFileSystemProvider(
    new EmbeddedFileProvider(...),
    new EmbeddedFileProvider(...));
```

Since the file provider is a hot path, using `IEnumerable<IFileProvider>` is not really appropriate. Not to mention that, semantically, it's not immediately clear that the order matters when using `IEnumerable`. IMHO, an array or a list is much more appropriate in this case. That said, nothing prevents you from adding an `IEnumerable<>` overload that calls `ToArray()` or `ToList()` to avoid allocating enumerators in `GetFileInfo`/`GetDirectoryContents`.

davidfowl replied:
> A params overload definitely makes sense, because that's how people will use this combined thingy:

Not really. Just make an array if you need that:

```C#
new CombinedFileSystemProvider(new[] {
    new EmbeddedFileProvider(...),
    new EmbeddedFileProvider(...) });
```

I'd keep the params array overload for a static factory method instead of baking it into the ctor.


Converting it to an array or list is more appropriate than enforcing an array be passed in.

mgrosperrin replied:
Hi @davidfowl, is it a general purpose to avoid params, or just in this situation ?

davidfowl commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,193 @@
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            // gets the content of all directories and merged them
+            var existingDirectoryContents = new List<IDirectoryContents>();
```
Instead of enumerating up front it might make sense to have the `CombinedDirectoryContents` take both file providers and do this in a more lazy matter (possible using yield or a custom iterator).

PinpointTownes replied:
https://github.com/aspnet/FileSystem/pull/142#discussion_r45621235

That said, @mgrosperrin is right: the impact will be imperceptible when using `PhysicalFileProvider`, because it immediately stores the items returned by `EnumerateFileSystemInfos` in a new list when `GetDirectoryContents` is called.

That's definitely a surprising behavior, specially since it uses `EnumerableDirectoryContents`, that would allow using an iterator (or easier, a good `Select` LINQy thingy :smile:)

https://github.com/aspnet/FileSystem/blob/dev/src/Microsoft.AspNet.FileProviders.Physical/PhysicalFileProvider.cs#L146-L163

VaclavElias commented:
Just a question. Is this provider going to allow to have Views in Class Library? If yes, how can we use it? If not, please ignore my question :). Thanks.

MartinJohns commented:
@VaclavElias You can do that already.

```C#
services.Configure<RazorViewEngineOptions>(options =>
{
    options.FileProvider = yourFileProvider;
});
```

By the default the used file provider is a `PhysicalFileProvider , pointing to your views directory. You can replace this with a custom `IFileProvider`, e.g. the `EmbeddedFileProvider` to support views embedded in class libraries.

The here proposed `CombinedFileProvider` approach let's you use a `PhysicalFileProvider` as a first option, and the `EmbeddedFileProvider` as a second option. This would use views present in the local file system if they exist, with a fall back to the views embedded in the assembly.

VaclavElias commented:
Thanks @MartinJohns making it clear for me!

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,193 @@
+
+            public CombinedFileChangeToken(List<IChangeToken> changeTokens)
+            {
+                _changeTokens = changeTokens ?? new List<IChangeToken>();
```
nit: looks like `changeTokens` cannot be null when used by `Watch`. I'd personally removed the null coalescing operator :smile:

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,193 @@
+            public IDisposable RegisterChangeCallback(Action<object> callback, object state)
+            {
+                var disposables = new List<IDisposable>();
+                foreach (var changeToken in _changeTokens)
```
Note: when using a list, the compiler doesn't apply any optimization. If you want to avoid the enumerator allocation here, you have to use a good old `for` loop.

PinpointTownes commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,193 @@
+                {
+                    if (_disposables != null)
+                    {
+                        foreach (var disposable in _disposables)
```
Same remark :+1:

MartinJohns commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,193 @@
+            return combinedFileChangeToken;
+        }
+
+        private class CombinedDirectoryContents : IDirectoryContents
```
I'm perhaps alone with this opinion and I've seen this in the other FileSystem projects already, but can't we stop with these nested classes and just make it an internal class within an own file?

mgrosperrin replied:
OK for me.
I could submit a new PR in order to do it after this one to un-nest these classes in all projects.

pranavkm replied:
You could start with this - just make these types public.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+        /// <param name="fileProviders"></param>
+        public CombinedFileProvider(params IFileProvider[] fileProviders)
+        {
+            _fileProviders = fileProviders;
```
`fileProviders` needs a null check.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+            foreach (var fileProvider in _fileProviders)
+            {
+                var changeToken = fileProvider.Watch(pattern);
+                if (changeToken != null && changeToken.ActiveChangeCallbacks)
```
Why would it only look at `ActiveChangeCallbacks`?

mgrosperrin replied:
This is the only way I found to determine if a `IChangeToken` is "real" and not a`NoopChangeToken`. I can combine all change tokens, but it will add some overload.

mgrosperrin replied:
In fact, I should not combined `NoopChangeToken` as its `RegisterChangeCallback` method always raises a `NotSupportedException`.

pranavkm replied:
This would miss any change tokens that passively expire. My guess is you might have to register it but only set up change callbacks on ones that support active callbacks.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+
+            public CombinedDirectoryContents(IFileProvider[] fileProviders, string subpath)
+            {
+                _directoriesContents =new Lazy<List<IDirectoryContents>>(() =>
```
Why make this deferred?

mgrosperrin replied:
This is in response to @davidfowl feedback: https://github.com/aspnet/FileSystem/pull/142#discussion_r45708539

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+                get { return _changeTokens.Any(_ => _.ActiveChangeCallbacks); }
+            }
+
+            private class Disposables : IDisposable
```
`CombinedDisposable` to maintain the naming pattern.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+
+            public bool HasChanged
+            {
+                get { return _changeTokens.Any(_ => _.HasChanged); }
```
Don't use underscores for things you're using: 
`_changeTokens.Any(token => token.HasChanged)`. Same comment for `ActiveChangeCallbacks`

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+                }
+                public void Dispose()
+                {
+                    if (_disposables != null)
```
Isn't this guaranteed to be non null?

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+    /// Looks up files using embedded resources in the specified assembly.
+    /// This file provider is case sensitive.
+    /// </summary>
+    public class CombinedFileProvider : IFileProvider
```
Should this be named `CompositeFileProvider`?

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs:
```diff
@@ -0,0 +1,205 @@
+namespace Microsoft.AspNet.FileProviders
+{
+    /// <summary>
+    /// Looks up files using embedded resources in the specified assembly.
```
Summary is incorrrect.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+        public void GetFileInfo_ReturnsNotFoundFileInfo_IfNoFileProviderSpecified()
+        {
+            // Arrange
+            var provider = new CombinedFileProvider();
```
I would probably have it throw if there are fewer than 2 elements registered.

mgrosperrin replied:
I'm not sure that imposing a minimum number of providers is a good thing (it is not in one of my use case: I want the list to be configurable, and the configuration can be to take only one source).

pranavkm replied:
It just feels weird when the CombinedFileProvider doesn't actually combine anything. Plus it's more work to cover these edge cases in tests

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+            var fileInfo = provider.GetFileInfo(fileName);
+
+            // Assert
+            Assert.NotNull(fileInfo);
```
You could replace the three statements with `Assert.Same(expectedFileInfo, fileInfo);`

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+        public void GetDirectoryContents_ReturnsCombinaisionOFFiles()
+        {
+            // Arrange
+            IFileInfo file1 = new MockFileInfo("File1"),
```
Use `var` and declare these as separate statements (instead of using commas)

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+                new MockFileProvider(
+                    file1,
+                    file2
+                    ),
```
Formatting: put these parentheses on the previous line e.g. `file2),`

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+            Assert.NotNull(files);
+            Assert.True(files.Exists);
+            Assert.Collection(files.OrderBy(f => f.Name, StringComparer.Ordinal),
+                file => Assert.Equal(file1, file),
```
Use `Assert.Same` for reference equality.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+        }
+
+        [Fact]
+        public void GetDirectoryContents_ReturnsCombinaisionOFFiles_WhenSomeFileProviderRetunsNoContent()
```
Typo: GetDirectoryContents_ReturnsCombinationOfFiles_WhenSomeFileProviderRetunsNoContent

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/project.json:
```diff
@@ -0,0 +1,29 @@
+    "dotnet5.4": {
+      "dependencies": {
+        "System.Collections": "4.0.11-*",
+        "System.Reflection": "4.0.10-*",
```
You might not need these.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockChangeToken.cs:
```diff
@@ -0,0 +1,57 @@
+
+namespace Microsoft.AspNet.FileProviders.Combined.Tests.TestUtility
+{
+    internal class MockChangeToken : IChangeToken
```
Could you rename this to `TestChangeToken`? `MockChangeToken` sounds like it's using `Moq`

mgrosperrin replied:
I have named it as is to be consistent with the naming in the FileSystemGlobbing test project. But I can change it :smile:

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockChangeToken.cs:
```diff
@@ -0,0 +1,57 @@
+﻿using Microsoft.Extensions.Primitives;
```
Could you add the license header to the file. Also sort usings so that `System` namespaces appear first.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.FileProviders.Embedded.Tests
```
`namespace Microsoft.AspNet.FileProviders.Combined`

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockChangeToken.cs:
```diff
@@ -0,0 +1,57 @@
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.FileProviders.Combined.Tests.TestUtility
```
Could you move this file to the root of the test directory?

mgrosperrin replied:
As for the name, I have put it in this folder to be consistent with the FileSystemGlobbing test project.
Do I move the three files ?

pranavkm replied:
Yup. It's got one test file, no point burying it one level deep.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockChangeToken.cs:
```diff
@@ -0,0 +1,57 @@
+            get; set;
+        }
+
+        public int NbOfCallback
```
`public int CallbacksCount => _callbacks.Count`. Alternatively just expose the backing list as a property.

mgrosperrin replied:
Do I make the same change to all readonly properties (eg ActiveChangeCallbacks) ?

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/project.json:
```diff
@@ -0,0 +1,20 @@
+    "commands": {
+        "test": "xunit.runner.aspnet"
+    },
+    "resource": "File.txt;sub/**;Resources/**",
```
Shouldn't need this.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockFileProvider.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public IChangeToken Watch(string filter)
+        {
+            if(_changeTokens != null && _changeTokens.ContainsKey(filter))
```
I think it would be fine to null ref if we called watch without registering any changeTokens.

```C#
if (_changeTokens.TryGetValue(filter, out changeToken))
{
   return changeToken;
}
```

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockFileProvider.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public MockFileProvider(params KeyValuePair<string, IChangeToken>[] changeTokens)
+        {
+            _changeTokens = changeTokens.ToDictionary(changeToken => changeToken.Key,
```
Could you use vs default indentation:
```
 _changeTokens = changeTokens.ToDictionary(
      changeToken => changeToken.Key,
      changeToken => changeToken.Value,
      StringComparer.Ordinal);
```

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/TestUtility/MockFileProvider.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Microsoft.AspNet.FileProviders.Combined.Tests.TestUtility
+{
+    internal class MockFileProvider : IFileProvider
```
public

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+            var secondChangeToken = new MockChangeToken();
+            var thirdChangeToken = new MockChangeToken();
+            var provider = new CombinedFileProvider(
+                new MockFileProvider(new KeyValuePair<string, IChangeToken>("pattern", firstChangeToken),
```
Use VS default indentation : https://github.com/aspnet/FileSystem/pull/142/files#r45886930

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,252 @@
+            {
+                hasBeenCalled = true;
+                result = item;
+            }, null);
```
Could you add a test that makes sure the state flows to individual ChangeTokens?

pranavkm commented:
:watch:

mgrosperrin commented:
Hi,
I hope I have answer to all feedback.
The remaining points are (if I do not forget):
* The name of the provider: @pranavkm suggest `CompositeFileProvider`.
* The usage of `params` in the constructor: personnaly, I prefer the API with the `params` keyword, but @davidfowl doesn't seems to like it
* The check on the number of providers: @pranavkm suggest to impose to have at least two providers, but as I said, one of my use case is to combined the provider based on some configuration and I want to allow only one provider
* Should the implementation of `IDirectoryContents` and `IChangeToken` be public or internal ?
* Should I rename the fake implementation in the tests be renamed to `TestXXX`instead of `MockXXX` as in the FileSystemGlobbing test project ?
* Do I move these types to the root of the project instead of the `TestUtility` folder as in the FileSystemGlobbing test project ?
Thanks.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedDirectoryContents.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Microsoft.AspNet.FileProviders
+{
+    internal class CombinedDirectoryContents : IDirectoryContents
```
`public`

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedDirectoryContents.cs:
```diff
@@ -0,0 +1,74 @@
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.FileProviders
```
@Eilon should these go under `Microsoft.AspNet.FileProviders.Combined`? The other two file providers are also placed under this namespace.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedDirectoryContents.cs:
```diff
@@ -0,0 +1,74 @@
+                }
+                return directories;
+            }
+                );
```
Formatting: needs to go on the previous line.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedDirectoryContents.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public CombinedDirectoryContents(IFileProvider[] fileProviders, string subpath)
+        {
+            _directoriesContents =new Lazy<List<IDirectoryContents>>(() =>
```
Formatting (needs spaces after the assignment). Bunch of places that have the same issue.

pranavkm replied:
Could we possibly just do this on a first chance basis rather than use `Lazy`?  i.e. in 

```C#
public IEnumerator<IFileInfo> GetEnumerator()
{
    EnsureInitialized();
    return _files;
}

public bool Exists
{
   get
   {
      EnsureInitialized();
      return _exists;
  }
}

private void EnsureInitialized()
{
    if (_directoryContents == null)
    {
           // Do the stuff from the constructor here
     }
}

```

We aren't dealing with concurrency here, so just deferring evaluation should suffice.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedDisposable.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            _disposables = disposables;
+        }
+        public void Dispose()
```
Need a blank line prior to this.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedDisposable.cs:
```diff
@@ -0,0 +1,25 @@
+        }
+        public void Dispose()
+        {
+            for (int i = 0; i < _disposables.Count; i++)
```
`var`

pranavkm replied:
Couple of other places with the same issue.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Combined/CombinedFileChangeToken.cs:
```diff
@@ -0,0 +1,42 @@
+            return new CombinedDisposable(disposables);
+        }
+
+        public bool HasChanged
```
You could rewrite this as 
`public bool HasChanged => _changeTokens.Any(token => token.HasChanged)`
Just a suggestion since it's not part of our guideline

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,250 @@
+                new MockFileProvider(
+                    new MockFileInfo("FileA"),
+                    new MockFileInfo("FileB")
+                    ),
```
Formatting. Parentheses on previous lines

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,250 @@
+        public void GetDirectoryContents_ReturnsCombinaitionOFFiles_WhenSomeFileProviderRetunsNoContent()
+        {
+            // Arrange
+            IFileInfo folderAFile1 = new MockFileInfo("FolderA/File1"),
```
`var`s on separate statements.

pranavkm commented on test/Microsoft.AspNet.FileProviders.Combined.Tests/CombinedFileProviderTests.cs:
```diff
@@ -0,0 +1,250 @@
+            thirdChangeToken.HasChanged = false;
+
+            // Register callback
+            Assert.Equal(0, firstChangeToken.Callsback.Count);
```
Could you split these into two tests? One's dealing with `HasChanged` and the other with callbacks.

pranavkm commented:
:watch:

pranavkm commented:
cc @Eilon for `CompositeFileProvider` versus `CombinedFileProvider`. Mvc has the habit of using `Composite` so it sounded like it would be a familiar pattern for users.

Eilon commented:
Yeah `Composite` sounds better to me for this.

mgrosperrin commented:
Hi,
is there anything else I can do to respond to your remarks?

pranavkm commented:
Sorry, GitHub doesn't send an email when you push an update. A comment saying you had it updated would be great (for future use). I'll have a look in a bit.

mgrosperrin commented:
Oh OK, didn't know that :smile: 
I'll know for the next time.

muratg commented:
@pranavkm Could you merge once it's good?

@mgrosperrin Thanks for the contribution!

pranavkm commented on src/Microsoft.AspNet.FileProviders.Composite/CompositeDirectoryContents.cs:
```diff
@@ -0,0 +1,83 @@
+
+namespace Microsoft.AspNet.FileProviders
+{
+    internal class CompositeDirectoryContents : IDirectoryContents
```
Could you make these types public? No reason to have them be internal.

pranavkm commented on src/Microsoft.AspNet.FileProviders.Composite/CompositeDirectoryContents.cs:
```diff
@@ -0,0 +1,83 @@
+
+        public CompositeDirectoryContents(IFileProvider[] fileProviders, string subpath)
+        {
+            _fileProviders = fileProviders;
```
This would need a null check (`throw new ArgumentNullException`)

pranavkm commented on src/Microsoft.AspNet.FileProviders.Composite/CompositeDirectoryContents.cs:
```diff
@@ -0,0 +1,83 @@
+        private bool _exists;
+        private List<IDirectoryContents> _directories;
+
+        public CompositeDirectoryContents(IFileProvider[] fileProviders, string subpath)
```
`IList<IFileProvider>`

pranavkm commented on src/Microsoft.AspNet.FileProviders.Composite/CompositeFileProvider.cs:
```diff
@@ -0,0 +1,85 @@
+        /// Initializes a new instance of the <see cref="CompositeFileProvider" /> class using a list of file provider.
+        /// </summary>
+        /// <param name="fileProviders"></param>
+        public CompositeFileProvider(params IFileProvider[] fileProviders)
```
Did we settle on using `params`? I thought @davidfowl wanted to use an `IEnumerable` here?

pranavkm commented:
Looks good. If you could get the last few changes.

MartinJohns commented:
This might be a crazy idea given the state of so much code in the repositories.. But can't we start to comment the code? Add some documentation?

mgrosperrin commented:
Hi,
I have made the changes.
For the constructor of `CompositeFileProvider`, I have added an overload with an `IEnumerable` and kept the current one.

pranavkm commented:
I'll get this merged in soon.

pranavkm commented:
Merged in 15c9f1d229b000d9b2472b1fcfd22120a116669e

pranavkm commented:
Thanks for the PR!

mgrosperrin commented:
You're welcome!
Thanks for your reviews.
